### PR TITLE
[ENG-164] New api registration workflow

### PR DESCRIPTION
**Pull request by corbinSanders**
## Purpose

Creating GET, POST endpoints at v2/draft_registrations and  PATCH, DELETE endpoints at v2/draft_registrations/<draft_id>.

## Changes

Pulling in Uday's osf models. Pulling in api/draft_registrations/, as well as all api_tests/draft_registrations tests.

## QA Notes

Ensure the above endpoints are functional

## Documentation

Document the new draft_registrations endpoints.
## Side Effects

Because of the close tie ins, this also includes the v2/registrations endpoints.

## Ticket

https://openscience.atlassian.net/browse/ENG-164

### CODE REVIEW

File: api_tests/users/views/test_user_draft_registration_list.py [hunk 1]
```diff
@@ -66,17 +66,17 @@ def test_view_permissions(
...
 
-        #   test_read_only_contributor_cannot_view_draft_list
+        #   test_read_only_contributor_can_view_draft_list
         res = app.get(
             url_draft_registrations,
             auth=user_read_contrib.auth)
-        assert len(res.json['data']) == 0
+        assert len(res.json['data']) == 1
 
         #   test_read_write_contributor_cannot_view_draft_list
```
Comment (UdayVarkhedkar) [inline api_tests/users/views/test_user_draft_registration_list.py]: Should this be `# test_read_write_contributor_can_view_draft_list`?

File: api/users/views.py [hunk 2]
```diff
@@ -496,21 +496,21 @@ class UserDraftRegistrations(JSONAPIBaseView, generics.ListAPIView, UserMixin):
...
-            Q(registered_node__isnull=True) |
-            Q(registered_node__is_deleted=True),
-            branched_from__in=list(node_qs),
-            deleted__isnull=True,
-        )
+        requested_user = self.get_user()
+        requesting_user = self.request.user
+        if requesting_user.is_anonymous:
+            # User making requests is unauthenticated so they can't see any Drafts
+            raise DraftRegistration.objects.none()
```
Comment (Johnetordoff) [inline api/users/views.py]: drive-by: raise? surely you mean return, also this should be hit by a unit test.

File: osf_tests/test_migration_sql.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        draft_reg.save()
+        assert(len(draft_reg.group_objects))
+
+    @pytest.mark.django_db
+    def test_drop_draft_reg_group_object_permission_table(self):
+        draft_registration = factories.DraftRegistrationFactory()
+        draft_registration.save()
+        draft_reg_group_obj_perm = DraftRegistrationGroupObjectPermission.objects.filter(content_object=draft_registration)[0]
+        draft_reg_group_obj_perm_id = draft_reg_group_obj_perm.id
+        assert(draft_reg_group_obj_perm)
```
Comment (Johnetordoff) [inline osf_tests/test_migration_sql.py]: Drive-by: This is pretty bizarre syntax here, if the previous line doesn't throw an error, this will pretty much always pass. Putting parenthesis around an assert like this isn't done anywhere else in the code base and should probably be avoided unless I'm missing something.

File: osf_tests/test_user.py [hunk 4]
```diff
@@ -387,6 +399,76 @@ def test_merge_preprints(self, user):
         assert not preprint_five.has_permission(user2, permissions.READ)
         assert not preprint_five.is_contributor(user2)
 
+    def test_merge_drafts(self, user):
```
Comment (Johnetordoff) [inline osf_tests/test_user.py]: I know this is consistent with other tests in this file, but consider breaking this up. Also the fact the `user2` is designated using the numeral, while `draft_two` is spelled out bothers me more then it probably should.

File: website/project/views/contributor.py [hunk 5]
```diff
@@ -541,8 +541,8 @@ def notify_added_contributor(node, contributor, auth=None, throttle=None, email_
 
     throttle = throttle or settings.CONTRIBUTOR_ADDED_EMAIL_THROTTLE
     # Email users for projects, or for components where they are not contributors on the parent node.
-    if contributor.is_registered and (isinstance(node, Preprint) or
-            (not node.parent_node or (node.parent_node and not node.parent_node.is_contributor(contributor)))):
+    if contributor.is_registered and ((isinstance(node, Preprint) or (isinstance(node, DraftRegistration)) or
```
Comment (Johnetordoff) [inline website/project/views/contributor.py]: You can shorten this to `isinstance(node, (Preprint, DraftRegistration))` for example `isinstance('hi!', (int, str)) == True`

File: api/draft_nodes/permissions.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+class ContributorOnDraftRegistration(IsContributorOrAdminContributor):
+    """
+    DraftNodes are hidden entities - only used to store files for DraftRegistrations.
+    Therefore, permission checking for DraftNodes will be done on the DraftRegistration.
+    """
+
+    acceptable_models = (DraftRegistration,)
+
+    def has_object_permission(self, request, view, obj):
+        if isinstance(obj, (DraftNode)):
```
Comment (Johnetordoff) [inline api/draft_nodes/permissions.py]: unnecessary parenthesis here

File: api/nodes/serializers.py [hunk 7]
```diff
@@ -140,6 +116,27 @@ def to_internal_value(self, data):
...
+    Used for both node_license field and license relationship.
+    Prevents license from having to be fetched 2x per node.
+    """
+    license_context = serializer.context.get('licenses', {})
+    if license_context and node._id in license_context:
+        return license_context.get(node._id)
+    else:
+        license = node.license
+        if license_context:
+            license_context[node._id] = {}
```
Comment (Johnetordoff) [inline api/nodes/serializers.py]: This line serves no purpose considering the following line.

**Comment by UdayVarkhedkar** [comment]
Based off of local testing, creating a new draft_registration (either given a pre-existing node or when creating a new draft_node) with metadata, does not apply the provided metadata in the POST body. Metadata changes are applied when PATCHing the DraftRegistrationDetail endpoint. Based off of `api.draft_registration.serializers`, it looks like metadata fields are being serialized but aren't actually used  when the draft_registration is created.

File: osf_tests/test_migration_sql.py (hunk 3 above)
Comment (UdayVarkhedkar) [reply osf_tests/test_migration_sql.py]: This is a test that I wrote for the DraftNode models/migrations - you're right and I'll fix it in this PR (https://github.com/CenterForOpenScience/osf.io/pull/9229)

**Comment by UdayVarkhedkar** [comment]
Also, it looks like you need to add this block to `EXTERNAL_EMBER_APPS` in `website.settings.local-dist.py`
```
    'registries': {
        'server': 'http://{}:4202/'.format(EMBER_DOMAIN),
        'path': '/registries/'
    },
```

**Comment by aaxelb** [comment]
subsumed by #9278
